force_wfmanager/left_side_pane/workflow_model_view.py

```diff
@@ -100,19 +101,57 @@ def remove_entity(self, entity):
                            " workflow".format(type(entity).__name__))
             )
 
-    def _get_mco_representation(self):
+    @on_trait_change('model.mco')
+    def update_mco_representation(self):
         if self.model.mco is not None:
-            return [MCOModelView(model=self.model.mco)]
+            self.mco_representation = [MCOModelView(model=self.model.mco)]
         else:
-            return []
-
-    def _get_data_sources_representation(self):
-        return [DataSourceModelView(model=data_source)
-                for data_source in self.model.data_sources]
-
-    def _get_kpi_calculators_representation(self):
-        return [KPICalculatorModelView(model=kpi_calculator)
-                for kpi_calculator in self.model.kpi_calculators]
+            self.mco_representation = []
+
+    @on_trait_change('model.data_sources[]')
+    def update_data_sources_representation(self):
+        self.data_sources_representation = [
+            DataSourceModelView(
+                model=data_source,
+                available_variables=self.mco_parameters_names)
+            for data_source in self.model.data_sources]
+
+    @on_trait_change('model.kpi_calculators[]')
+    def update_kpi_calculators_representation(self):
+        available_variables = self.mco_parameters_names + \
+            self.data_source_outputs_names
+        self.kpi_calculators_representation = [
+            KPICalculatorModelView(
+                model=kpi_calculator,
+                available_variables=available_variables)
+            for kpi_calculator in self.model.kpi_calculators]
+
+    @on_trait_change('mco_representation.mco_parameters_names[]')
+    def update_mco_parameters_names(self):
+        if len(self.mco_representation) != 0:
+            self.mco_parameters_names = \
+                self.mco_representation[0].mco_parameters_names
+        else:
+            self.mco_parameters_names = []
+
+    @on_trait_change('mco_parameters_names[]')
+    def update_data_sources_inputs(self):
+        for data_source_mv in self.data_sources_representation:
```

Comment:
This scales badly, every time the user will type a new letter in a name of an MCO parameter, it will trigger this and be propagated everywhere.